# Introduction to Pandas I
Welcome to the Pandas I. In this lesson we will be covering: 
- **Creating, reading and writing DataFrames**
- **Indexing in Pandas**
- **Mapping and Summarizing with Pandas**

The lab for Lesson 3 will consist of the nine exercises that you will find throughtout the notebook. 

For this exercise we wil be using the Titanic Survival Dataset from Kaggle. We will perform various tranformations, edits and exploration. 

These are the various features present for each passenger on the ship:
- **Survived**: Outcome of survival (0 = No; 1 = Yes)
- **Pclass**: Socio-economic class (1 = Upper class; 2 = Middle class; 3 = Lower class)
- **Name**: Name of passenger
- **Sex**: Sex of the passenger
- **Age**: Age of the passenger (Some entries contain `?`)
- **SibSp**: Number of siblings and spouses of the passenger aboard
- **Parch**: Number of parents and children of the passenger aboard
- **Ticket**: Ticket number of the passenger
- **Fare**: Fare paid by the passenger
- **Cabin** Cabin number of the passenger (Some entries contain `?`)
- **Embarked**: Port of embarkation of the passenger (C = Cherbourg; Q = Queenstown; S = Southampton)
- **Home.Dest**: Home / Destination

Let's start with importing some of the libraries we will be using.

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt

### Why Pandas?
Pandas is the go-to python library to use for any data scientist or machine learning engineer. It allows for easy data access, easy data manipulation, and it's free! Let's start off by reading the titanic survival dataset. 

## Creating and Reading Data

### Reading in Data

In [2]:
# Read in the titanic survival dataset 
titanic_data = pd.read_csv('titanic_data.csv')
titanic_data

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex     age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female      29      0      0   24160  211.3375       B5        S    2   
1       male  0.9167      1      2  113781    151.55  C22 C26        S   11   
2     female       2      1      2  113781    151.55  C22 C26        S    ?   
3       male      30      1      2  113781    151.55  C22 C26        S    ?   
4     female      25      1      2  113781    151.55  C22 C26        S    ?   
...      ...     ...    ...    ...     ...       ...      ...      ...  ...   
1304  female    14.5      1      0    2665   14.4542        ?        C    ?   
1305  female       ?      1      0    2665   14.4542        ?        C    ?   
1306    male    26.5      0      0    2656     7.225        ?        C    ?   
1307    male      27      0      0    2670     7.225        ?        C    ?   
1308    male      29      0      0  315082     7.875        ?        S    ?   

     body                        home.dest  
0       ?                     St Louis, MO  
1       ?  Montreal, PQ / Chesterville, ON  
2       ?  Montreal, PQ / Chesterville, ON  
3     135  Montreal, PQ / Chesterville, ON  
4       ?  Montreal, PQ / Chesterville, ON  
...   ...                              ...  
1304  328                                ?  
1305    ?                                ?  
1306  304                                ?  
1307    ?                                ?  
1308    ?                                ?  

[1309 rows x 14 columns]

Pandas has various options to read data. If you are curious, you can read about them here: [Pandas Input/Output](https://pandas.pydata.org/pandas-docs/stable/reference/io.html). For the majority of the exercises, we will be using `read_csv`. If you would like to see all the options for read_csv, [here is the documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html#pandas.read_csv). 
A CSV file is a table of values seperated by commas (CSV =  Comma Seperated Values). The file looks like:

data1,data2,data3,...,datan

The data is read into a DataFrame. A DataFrame, as you can see from above, is a table. Every entry corresponds to a row and column.

### Pandas Series 
A Series is a sequence of data. A dataframe is a table and a series is a list. What's the difference between a Series and a DataFrame? A Dataframe is a table and a Series would be just a single column of data. DataFrames have column names and Series do not. 

In [3]:
# Lets take a look at A Series within the titanic survival dataset. 
titanic_series_data = titanic_data['survived']
titanic_series_data

0       1
1       1
2       0
3       0
4       0
       ..
1304    0
1305    0
1306    0
1307    0
1308    0
Name: survived, Length: 1309, dtype: int64

In [4]:
# EXERCISE 1
# choose a column and create your own series below

### Creating Data Using Pandas
If you wanted to create your own dataset using pandas you can use the DataFrame method. The DataFrame method can take in data in various formats and creates a dataframe object that can be used to store data. In the example below we are creating a new DataFrame. Note the syntax required to define a header (Survivor_#) versus the data (Name). 

In [5]:
my_data_frame = pd.DataFrame({"Survivor_1":["Nelson"],"Survivor_2":["Monica"],"Survivor_3":["Marlon"]})
my_data_frame

Survivor_1 Survivor_2 Survivor_3
0     Nelson     Monica     Marlon

In [6]:
# Think: how would you add a second row to the data above?

In [7]:
# EXERCISE 2
# Create your own DataFrame. Name your DataFrame your_name_df, and assign column values with your first and last name
# and the values will be the length of your first and last name


## Selecting, Indexing and Assigning
Selecting, indexing, and assigning data will be your bread and butter of data analysis. These tools will allow you to select aspects of your data quickly and efficiently to analyze. 

### Native Accessing 
Let's start by selecting a column of data. There are two ways to do this, either by accessing the column using the dot operator or by referencing the column explicitly.

In [75]:
# Selecting one column data using the dot operator
titanic_data.ticket

0        24160
1       113781
2       113781
3       113781
4       113781
         ...  
1304      2665
1305      2665
1306      2656
1307      2670
1308    315082
Name: ticket, Length: 1309, dtype: object

In [76]:
# Selecting one column by referencing the column explicitly
titanic_data['ticket']

0        24160
1       113781
2       113781
3       113781
4       113781
         ...  
1304      2665
1305      2665
1306      2656
1307      2670
1308    315082
Name: ticket, Length: 1309, dtype: object

There is no real advantage of which to use. They both perform the same function, but when the column is referenced explicitly, you can use column names with special characters such as: "Test Data", "First.Last".

In [10]:
# EXERCISE 3
# Select the home.dest column fromt the titanic survival dataset and print it to the screen. 


### Indexing in Pandas

Now let's take a look at the first value of a column. Let's reference the column explicitly using square brackets:

In [78]:
titanic_data['ticket'][1]

'113781'

We were able to access the second row *(zero-based indexing)* of data from the ticket column by referencing the column explicity and by using square brackets. Now the method that is typically used in industry, and is the recommended way of accessing data, is iloc. 

Lets use iloc below and access the same data as we did above

In [79]:
# Using iloc by natively acessing a dataframe column. 
titanic_data['ticket'].iloc[1]

'113781'

We used `iloc` above to access the second value, `iloc[1]`, from the column ticket.

`iloc` accesses data by using row-first, column-second.

- iloc[row,column]

If we wanted to perform the operation above without natively accessing the column, we can use `iloc` in the following way 

In [80]:
# Using iloc to access the first row, from the seventh column
titanic_data.iloc[1,7]

'113781'

Now let's show more ways we can use `iloc`. 

In [81]:
# Using iloc lets access all the rows in the first column
titanic_data.iloc[:,0]

0       1
1       1
2       1
3       1
4       1
       ..
1304    3
1305    3
1306    3
1307    3
1308    3
Name: pclass, Length: 1309, dtype: int64

In [82]:
# Using iloc to access the first three rows, from the first column
titanic_data.iloc[:3,0]

0    1
1    1
2    1
Name: pclass, dtype: int64

In [83]:
# Using iloc to access the last 5 rows from the first column
titanic_data.iloc[-5:,0]

1304    3
1305    3
1306    3
1307    3
1308    3
Name: pclass, dtype: int64

`iloc` is not the only method that can be used to access data. The `loc` method can also be used similarly, but there is a difference between the two. `loc` is used to access data using label-based selection. So when we use `loc` we will include a label to use as shown below.

In [84]:
#Using loc to access the first three rows from the age column
titanic_data.loc[0:2,'age']

0    29.0000
1     0.9167
2     2.0000
Name: age, dtype: float64

In [85]:
# EXERCISE 4
# Describe the difference between iloc and loc, using your own words

#Answer here

Now let's give it a try on our own. Perform the following operations using iloc:

In [86]:
# EXERCISE 5
# Access the 100th row from the cabin column
cabin_value = 
print("The cabin value of the 100th row is {}".format(cabin_value))


# Access rows 300-310 from the ticket column 
ticket_values = 
print("Tickets values from indicies 300-310 are: {}".format(ticket_values))


# Access the the second row, and the last 5 columns from the Titanic Data set 
five_columns_value = 
print("The last five columns contain the following values in the first row: {}".format(five_columns_value))

SyntaxError: invalid syntax (<ipython-input-86-0d8a81aaf330>, line 3)

### Manipulating the Index
If you have not noticed yet, everytime we print a pandas DataFrame to the screen, there is a column at the beginning that starts at 0. This column is the index column and it is what Pandas uses as a refrence point for every DataFrame. Let's take a look at how to access the index column.  

In [87]:
titanic_data.index

RangeIndex(start=0, stop=1309, step=1)

The information above details that the index starts at 0, increments by 1 and stops at 1309.

The index will not always be in order, or use numbers as increments. You can use label values as index, for which you can later use to help you make graphs, or use `loc` to access specific data. Below we will change the index to sex, and show you how you can change the index value in a Pandas DataFrame

In [88]:
#changing the index
titanic_data_new_index = titanic_data.set_index("sex")
titanic_data_new_index

pclass  survived                                             name  \
sex                                                                         
female       1         1                    Allen, Miss. Elisabeth Walton   
male         1         1                   Allison, Master. Hudson Trevor   
female       1         0                     Allison, Miss. Helen Loraine   
male         1         0             Allison, Mr. Hudson Joshua Creighton   
female       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...        ...       ...                                              ...   
female       3         0                             Zabour, Miss. Hileni   
female       3         0                            Zabour, Miss. Thamine   
male         3         0                        Zakarian, Mr. Mapriededer   
male         3         0                              Zakarian, Mr. Ortin   
male         3         0                               Zimmerman, Mr. Leo   

            age  sibsp  parch  ticket      fare    cabin embarked boat body  \
sex                                                                           
female  29.0000      0      0   24160  211.3375       B5       UK    2    ?   
male     0.9167      1      2  113781    151.55  C22 C26       UK   11    ?   
female   2.0000      1      2  113781    151.55  C22 C26       UK    ?    ?   
male    30.0000      1      2  113781    151.55  C22 C26       UK    ?  135   
female  25.0000      1      2  113781    151.55  C22 C26       UK    ?    ?   
...         ...    ...    ...     ...       ...      ...      ...  ...  ...   
female  14.5000      1      0    2665   14.4542        ?       FR    ?  328   
female      NaN      1      0    2665   14.4542        ?       FR    ?    ?   
male    26.5000      0      0    2656     7.225        ?       FR    ?  304   
male    27.0000      0      0    2670     7.225        ?       FR    ?    ?   
male    29.0000      0      0  315082     7.875        ?       UK    ?    ?   

                              home.dest Ship_Name  last_name   first_name  
sex                                                                        
female                     St Louis, MO   Titanic      Allen    Elisabeth  
male    Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
female  Montreal, PQ / Chesterville, ON   Titanic    Allison        Helen  
male    Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
female  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
...                                 ...       ...        ...          ...  
female                                ?   Titanic     Zabour       Hileni  
female                                ?   Titanic     Zabour      Thamine  
male                                  ?   Titanic   Zakarian  Mapriededer  
male                                  ?   Titanic   Zakarian        Ortin  
male                                  ?   Titanic  Zimmerman          Leo  

[1309 rows x 16 columns]

### Assigning data
To assign new data to an already existing DataFrame is very simple. Let's add our own column to the titanic dataset:

In [89]:
# Add a new column named Ship_Name and populate it with the value Titanic
titanic_data['Ship_Name'] = "Titanic"
titanic_data

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5       UK    2   
1       male   0.9167      1      2  113781    151.55  C22 C26       UK   11   
2     female   2.0000      1      2  113781    151.55  C22 C26       UK    ?   
3       male  30.0000      1      2  113781    151.55  C22 C26       UK    ?   
4     female  25.0000      1      2  113781    151.55  C22 C26       UK    ?   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542        ?       FR    ?   
1305  female      NaN      1      0    2665   14.4542        ?       FR    ?   
1306    male  26.5000      0      0    2656     7.225        ?       FR    ?   
1307    male  27.0000      0      0    2670     7.225        ?       FR    ?   
1308    male  29.0000      0      0  315082     7.875        ?       UK    ?   

     body                        home.dest Ship_Name  last_name   first_name  
0       ?                     St Louis, MO   Titanic      Allen    Elisabeth  
1       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
2       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison        Helen  
3     135  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
4       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
...   ...                              ...       ...        ...          ...  
1304  328                                ?   Titanic     Zabour       Hileni  
1305    ?                                ?   Titanic     Zabour      Thamine  
1306  304                                ?   Titanic   Zakarian  Mapriededer  
1307    ?                                ?   Titanic   Zakarian        Ortin  
1308    ?                                ?   Titanic  Zimmerman          Leo  

[1309 rows x 17 columns]

As you can see we created the column "Ship_Name" and assigned it a string value of "Titanic".

We can now combine some of what we have learned to create two new columns from the data we have. Using the name column, let's seperate the values into first and last name. The code below may seem new or a bit complicated, but don't worry, we will discuss what we are doing in the class.  
Run the cell below to create the new columns.

In [90]:
# Create the last name column
titanic_data_last_name = titanic_data
titanic_data_last_name['last_name'] = titanic_data['name'].str.split().str[0].replace(',','',regex=True)
titanic_data_last_name

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5       UK    2   
1       male   0.9167      1      2  113781    151.55  C22 C26       UK   11   
2     female   2.0000      1      2  113781    151.55  C22 C26       UK    ?   
3       male  30.0000      1      2  113781    151.55  C22 C26       UK    ?   
4     female  25.0000      1      2  113781    151.55  C22 C26       UK    ?   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542        ?       FR    ?   
1305  female      NaN      1      0    2665   14.4542        ?       FR    ?   
1306    male  26.5000      0      0    2656     7.225        ?       FR    ?   
1307    male  27.0000      0      0    2670     7.225        ?       FR    ?   
1308    male  29.0000      0      0  315082     7.875        ?       UK    ?   

     body                        home.dest Ship_Name  last_name   first_name  
0       ?                     St Louis, MO   Titanic      Allen    Elisabeth  
1       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
2       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison        Helen  
3     135  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
4       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
...   ...                              ...       ...        ...          ...  
1304  328                                ?   Titanic     Zabour       Hileni  
1305    ?                                ?   Titanic     Zabour      Thamine  
1306  304                                ?   Titanic   Zakarian  Mapriededer  
1307    ?                                ?   Titanic   Zakarian        Ortin  
1308    ?                                ?   Titanic  Zimmerman          Leo  

[1309 rows x 17 columns]

In [24]:
# Note that the split() method turns a string into a list, then you choose which value from the list you want to use. 

In [91]:
# Create the first name column
titanic_data_name = titanic_data_last_name
titanic_data_name['first_name'] = titanic_data['name'].str.split().str[2].replace(',','',regex=True)
titanic_data_name

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5       UK    2   
1       male   0.9167      1      2  113781    151.55  C22 C26       UK   11   
2     female   2.0000      1      2  113781    151.55  C22 C26       UK    ?   
3       male  30.0000      1      2  113781    151.55  C22 C26       UK    ?   
4     female  25.0000      1      2  113781    151.55  C22 C26       UK    ?   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542        ?       FR    ?   
1305  female      NaN      1      0    2665   14.4542        ?       FR    ?   
1306    male  26.5000      0      0    2656     7.225        ?       FR    ?   
1307    male  27.0000      0      0    2670     7.225        ?       FR    ?   
1308    male  29.0000      0      0  315082     7.875        ?       UK    ?   

     body                        home.dest Ship_Name  last_name   first_name  
0       ?                     St Louis, MO   Titanic      Allen    Elisabeth  
1       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
2       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison        Helen  
3     135  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
4       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
...   ...                              ...       ...        ...          ...  
1304  328                                ?   Titanic     Zabour       Hileni  
1305    ?                                ?   Titanic     Zabour      Thamine  
1306  304                                ?   Titanic   Zakarian  Mapriededer  
1307    ?                                ?   Titanic   Zakarian        Ortin  
1308    ?                                ?   Titanic  Zimmerman          Leo  

[1309 rows x 17 columns]

As you can see, we now have two new columns! first name and last_name. Let's try creating some columns of your own.

In [92]:
# EXERCISE 6 
# Create a new column called Deck_Location, and assign the value "Upper" to the the first half of the dataset, 
# and Lower to the second half of the dataset. Print the dataset once you are done.


### Conditional Selection
Now we have been using `iloc` to select rows and columns from a dataset. But what if you wanted to find values that matched a certain condition?  
This is where we can use conditional selection. In the following cells we will find values that meet conditions from our titanic dataset.

In [93]:
# Using conditional selection, let's find all the female passengers 
titanic_data.sex == 'female'

# Here, we're using the dot operator to access the 'sex' column and setting it equal to 'female'. 
# This will return True if the passenseg is a female and False if the passenger is a male

0        True
1       False
2        True
3       False
4        True
        ...  
1304     True
1305     True
1306    False
1307    False
1308    False
Name: sex, Length: 1309, dtype: bool

The output above gives us the index values of where our condition was either True or False. But what if we wanted to see all the date for female passengers instead the conditonal values? We can use `loc` to do this.

In [94]:
# Show the data for female passengers only
titanic_data.loc[titanic_data.sex == 'female']

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
2          1         0                     Allison, Miss. Helen Loraine   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
6          1         1                Andrews, Miss. Kornelia Theodosia   
8          1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)   
...      ...       ...                                              ...   
1286       3         1  Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1290       3         1                 Wilkes, Mrs. James (Ellen Needs)   
1300       3         1          Yasbeck, Mrs. Antoni (Selini Alexander)   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   

         sex   age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0      0      0   24160  211.3375       B5       UK    2   
2     female   2.0      1      2  113781    151.55  C22 C26       UK    ?   
4     female  25.0      1      2  113781    151.55  C22 C26       UK    ?   
6     female  63.0      1      0   13502   77.9583       D7       UK   10   
8     female  53.0      2      0   11769   51.4792     C101       UK    D   
...      ...   ...    ...    ...     ...       ...      ...      ...  ...   
1286  female  38.0      0      0    2688    7.2292        ?       FR    C   
1290  female  47.0      1      0  363272         7        ?       UK    ?   
1300  female  15.0      1      0    2659   14.4542        ?       FR    ?   
1304  female  14.5      1      0    2665   14.4542        ?       FR    ?   
1305  female   NaN      1      0    2665   14.4542        ?       FR    ?   

     body                        home.dest Ship_Name last_name first_name  
0       ?                     St Louis, MO   Titanic     Allen  Elisabeth  
2       ?  Montreal, PQ / Chesterville, ON   Titanic   Allison      Helen  
4       ?  Montreal, PQ / Chesterville, ON   Titanic   Allison     Hudson  
6       ?                       Hudson, NY   Titanic   Andrews   Kornelia  
8       ?              Bayside, Queens, NY   Titanic  Appleton     Edward  
...   ...                              ...       ...       ...        ...  
1286    ?                                ?   Titanic    Whabee     George  
1290    ?                                ?   Titanic    Wilkes      James  
1300    ?                                ?   Titanic   Yasbeck     Antoni  
1304  328                                ?   Titanic    Zabour     Hileni  
1305    ?                                ?   Titanic    Zabour    Thamine  

[466 rows x 17 columns]

We now have the data for all the female passengers using conditional selection.  
Let's try a few more:

In [95]:
# Access the data where for the female passengers that survived
titanic_data.loc[(titanic_data.sex == 'female') & (titanic_data.survived == 1)]

# Note that the notation above uses '&' instead of 'and'
# Using and (logical operator) asks python to check if the entire sex column is equal to 'female'
# & (bitwise operator) allows python to check individual values within the column

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
...      ...       ...                                                ...   
1260       3         1                            Turja, Miss. Anna Sofia   
1261       3         1                             Turkula, Mrs. (Hedwig)   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1300       3         1            Yasbeck, Mrs. Antoni (Selini Alexander)   

         sex   age  sibsp  parch    ticket      fare    cabin embarked boat  \
0     female  29.0      0      0     24160  211.3375       B5       UK    2   
6     female  63.0      1      0     13502   77.9583       D7       UK   10   
8     female  53.0      2      0     11769   51.4792     C101       UK    D   
11    female  18.0      1      0  PC 17757   227.525  C62 C64       FR    4   
12    female  24.0      0      0  PC 17477      69.3      B35       FR    9   
...      ...   ...    ...    ...       ...       ...      ...      ...  ...   
1260  female  18.0      0      0      4138    9.8417        ?       UK   15   
1261  female  63.0      0      0      4134    9.5875        ?       UK   15   
1286  female  38.0      0      0      2688    7.2292        ?       FR    C   
1290  female  47.0      1      0    363272         7        ?       UK    ?   
1300  female  15.0      1      0      2659   14.4542        ?       FR    ?   

     body            home.dest Ship_Name last_name first_name  
0       ?         St Louis, MO   Titanic     Allen  Elisabeth  
6       ?           Hudson, NY   Titanic   Andrews   Kornelia  
8       ?  Bayside, Queens, NY   Titanic  Appleton     Edward  
11      ?         New York, NY   Titanic     Astor       John  
12      ?        Paris, France   Titanic    Aubart   Leontine  
...   ...                  ...       ...       ...        ...  
1260    ?                    ?   Titanic     Turja       Anna  
1261    ?                    ?   Titanic   Turkula   (Hedwig)  
1286    ?                    ?   Titanic    Whabee     George  
1290    ?                    ?   Titanic    Wilkes      James  
1300    ?                    ?   Titanic   Yasbeck     Antoni  

[339 rows x 17 columns]

In [96]:
# Access the data where the passenger last name is either Allen or Allison 
titanic_data.loc[titanic_data.last_name.isin(['Allen','Allison'])]

# the 'isin()' method allows you to filter the dataset by selecting rows with a particular value(s)

pclass  survived                                             name  \
0         1         1                    Allen, Miss. Elisabeth Walton   
1         1         1                   Allison, Master. Hudson Trevor   
2         1         0                     Allison, Miss. Helen Loraine   
3         1         0             Allison, Mr. Hudson Joshua Creighton   
4         1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
618       3         0                         Allen, Mr. William Henry   

        sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0    female  29.0000      0      0   24160  211.3375       B5       UK    2   
1      male   0.9167      1      2  113781    151.55  C22 C26       UK   11   
2    female   2.0000      1      2  113781    151.55  C22 C26       UK    ?   
3      male  30.0000      1      2  113781    151.55  C22 C26       UK    ?   
4    female  25.0000      1      2  113781    151.55  C22 C26       UK    ?   
618    male  35.0000      0      0  373450      8.05        ?       UK    ?   

    body                                          home.dest Ship_Name  \
0      ?                                       St Louis, MO   Titanic   
1      ?                    Montreal, PQ / Chesterville, ON   Titanic   
2      ?                    Montreal, PQ / Chesterville, ON   Titanic   
3    135                    Montreal, PQ / Chesterville, ON   Titanic   
4      ?                    Montreal, PQ / Chesterville, ON   Titanic   
618    ?  Lower Clapton, Middlesex or Erdington, Birmingham   Titanic   

    last_name first_name  
0       Allen  Elisabeth  
1     Allison     Hudson  
2     Allison      Helen  
3     Allison     Hudson  
4     Allison     Hudson  
618     Allen    William

Let's try conditional selection on our own. 

In [97]:
# EXERCISE 7

# Access the values where the passenger survived and their first name was Elisabeth

In [32]:
# Access the values where the fare value equals "2665" and the passenger survived 

In [33]:
# Access the values where the fare values equals "113781" and the passenger survived

## Summary and Mapping

### Summary Functions 
Summary functions are used to quickly summarize your data. They are very handy tools when you are working with a new dataset and we will explore in a later lesson why that is important. 

Let's start by using using the `describe()` method. 

In [98]:
# Use the describe method to quickly summarize the titanic survival dataset
titanic_data.describe()

pclass     survived          age        sibsp        parch
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000
mean      2.294882     0.381971    29.881135     0.498854     0.385027
std       0.837836     0.486055    14.413500     1.041658     0.865560
min       1.000000     0.000000     0.166700     0.000000     0.000000
25%       2.000000     0.000000    21.000000     0.000000     0.000000
50%       3.000000     0.000000    28.000000     0.000000     0.000000
75%       3.000000     1.000000    39.000000     1.000000     0.000000
max       3.000000     1.000000    80.000000     8.000000     9.000000

The result of the `describe()` method are 8 values: 
- **count:** 
  count number of non-NA/null observations.
- **mean:**
  mean of the values.
- **std:**
  standard deviation of the observations.
- **min:**
  minimum of the values in the object.
- **max:**
  maximum of the values in the object.
- **25%-75%:**
  these values represent quartiles. We will go over what these mean in another lesson


If you want to only calculate one value, you can use the following:

In [99]:
# individual columns stats 
titanic_data.survived.mean()

0.3819709702062643

In [100]:
# EXERCISE 8

# Calculate the count of the survived column


#### The `unique()` and `value_counts()` methods 

The following methods allow you to quickly summarize a column:

- **unique:** This method will display all the unique values of your column. 
- **value_counts:** This method will display all the unique values and their counts. 

In [101]:
# Unique names
titanic_data.sex.unique()

array(['female', 'male'], dtype=object)

In [38]:
#value counts 
titanic_data.sex.value_counts()

male      843
female    466
Name: sex, dtype: int64

In [102]:
# EXERCISE 9

# Use the value_counts() method on the AGE column and print the results. 


### Mapping Functions
`map()` is a method that is very useful within Data Science and AI to alter or create new dipictions of data. The method takes a function and then maps that function to a set of values. This method can only be used with Series data and not a pandas dataframe.

Let's use both `map()` and `lambda` to normalize age values in the titanic dataset. 

To normilie the data, we will be "scaling" the values to be between 0 and 1 using the following formula:

$z_{nor}$ = (x - $x_{min}$) / ($x_{max}$ - $x_{min}$)

In [103]:
# First, we will use the code below to convert the age column to numeric values and get rid of the '?' in the data
titanic_data['age'] = pd.to_numeric(titanic_data['age'],errors='coerce')

# Find the max and min from the age column
max_age = titanic_data.age.max()
min_age = titanic_data.age.min()

# Use the Mapping Function
titanic_data.age.map(lambda p: (p - min_age)/(max_age - min_age))

0       0.361169
1       0.009395
2       0.022964
3       0.373695
4       0.311064
          ...   
1304    0.179540
1305         NaN
1306    0.329854
1307    0.336117
1308    0.361169
Name: age, Length: 1309, dtype: float64

Another useful method that we can use to alter data is the `apply()` method. This method is similar to the `map()` method, but it can be used with a pandas DataFrame.

Let's create a method that will convert the embarked column from the town of embarkation to the country of embarkation. 

In [104]:
def rename_embarked(row):
    if row.embarked == 'S':
        row.embarked = 'UK'
    elif row.embarked == 'C':
        row.embarked = 'FR'
    elif row.embarked == 'Q':
        row.embarked = 'IE'
    return row
    
titanic_data = titanic_data.apply(rename_embarked,axis='columns')
titanic_data

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5       UK    2   
1       male   0.9167      1      2  113781    151.55  C22 C26       UK   11   
2     female   2.0000      1      2  113781    151.55  C22 C26       UK    ?   
3       male  30.0000      1      2  113781    151.55  C22 C26       UK    ?   
4     female  25.0000      1      2  113781    151.55  C22 C26       UK    ?   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542        ?       FR    ?   
1305  female      NaN      1      0    2665   14.4542        ?       FR    ?   
1306    male  26.5000      0      0    2656     7.225        ?       FR    ?   
1307    male  27.0000      0      0    2670     7.225        ?       FR    ?   
1308    male  29.0000      0      0  315082     7.875        ?       UK    ?   

     body                        home.dest Ship_Name  last_name   first_name  
0       ?                     St Louis, MO   Titanic      Allen    Elisabeth  
1       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
2       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison        Helen  
3     135  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
4       ?  Montreal, PQ / Chesterville, ON   Titanic    Allison       Hudson  
...   ...                              ...       ...        ...          ...  
1304  328                                ?   Titanic     Zabour       Hileni  
1305    ?                                ?   Titanic     Zabour      Thamine  
1306  304                                ?   Titanic   Zakarian  Mapriededer  
1307    ?                                ?   Titanic   Zakarian        Ortin  
1308    ?                                ?   Titanic  Zimmerman          Leo  

[1309 rows x 17 columns]

In [73]:
# EXERCISE 9

# Replace the `?` value in the fare column with NAN and print the new fare column.
# Create your own mapping function for the fare column and print the new DataFrame.